### ISSUE TITLE
problem with save_data file writes

### ISSUE THREAD

**Issue by pavlis** [mspass-team/mspass#345]
Multiple things appear to be broken in the save_data method of Database when writing to files with "format == None".   That triggers a block of C code that writes the raw data with a binary fwrite to  a file.   The offender is this private functions:
```
@staticmethod
    def _save_data_to_dfile(
        mspass_object, dir, dfile, format=None, kill_on_failure=False
    ):
        """
        This is a private method used under the hood to save the sample data
        of atomic MsPASS data objects.  How the save happens is highly
        dependent upon the format argument.  When None, which is the
        default, the data are written to the file specified by dir and dfile
        using low-level C fwrite calls.  The function used always appends
        data to eof if the file already exists to allow accumation of data
        in "gather" files to reduce file name overhead in hpc systems.
        If format is anything else the function attempts to use one of
        the obspy formatted writers.  That means the format string must be
        one of the `supported formats <https://docs.obspy.org/packages/autogen/obspy.core.stream.Stream.write.html#supported-formats>`__ of ObsPy reader.
        Writing to a file can fail for any number of reasons.   write errors
        are trapped internally in this function any errors posted to elog of
        mspass_object.   If the kill_on_failure boolean is set true the
        function will call the kill method of the data function and the
        that datum will be marked dead.   That is not normally a good idea
        if there is any additional work to do on the data so the default
        for that parameter is false.
        Because we use a stream model for file storage be aware
        that insertion and deletion in a file are not possible.  If you
        need lots of editing functionality with files you should use the
        one file to one object model or (better yet) use gridfs storage.
        :param mspass_object: the target object.
        :type mspass_object: either :class:`mspasspy.ccore.seismic.TimeSeries` or :class:`mspasspy.ccore.seismic.Seismogram`
        :param dir: file directory.
        :type dir: :class:`str`
        :param dfile: file name.
        :type dfile: :class:`str`
        :param format: the format of the file. This can be one of the `supported formats <https://docs.obspy.org/packages/autogen/obspy.core.stream.Stream.write.html#supported-formats>`__ of ObsPy reader. By default (``None``), the format will be the binary waveform.
        :type format: :class:`str`
        :param kill_on_failure:  When true if an io error occurs the data object's
          kill method will be invoked.  When false (the default) io errors are
          logged and left set live.  (Note data already marked dead are return
          are ignored by this function. )
        :type kill_on_failure: boolean
        :return: Position of first data sample (foff) and the size of the saved chunk.
           If the input is flagged as dead return (-1,0)
        """
        if not isinstance(mspass_object, (TimeSeries, Seismogram)):
            raise TypeError("only TimeSeries and Seismogram are supported")
        if mspass_object.dead():
            return -1, 0

        if not format:
            try:
                # This function has overloading.  this might not work
                foff = _fwrite_to_file(mspass_object, dir, dfile)
            except MsPASSError as merr:
                mspass_object.elog.log_error(merr)
                if kill_on_failure:
                    mspass_object.kill()
            # we can compute the number of bytes written for this case
            if isinstance(mspass_object, TimeSeries):
                nbytes_written = 8 * mspass_object.npts
            else:
                # This can only be a Seismogram currently so this is not elif
                nbytes_written = 24 * mspass_object.npts
        else:
            fname = os.path.join(dir, dfile)
            os.makedirs(os.path.dirname(fname), exist_ok=True)
            with open(fname, mode="a+b") as fh:
                foff = fh.seek(0, 2)
                f_byte = io.BytesIO()
                if isinstance(mspass_object, TimeSeries):
                    mspass_object.toTrace().write(f_byte, format=format)
                elif isinstance(mspass_object, Seismogram):
                    mspass_object.toStream().write(f_byte, format=format)
                f_byte.seek(0)
                ub = f_byte.read()
                fh.write(ub)
                nbytes_written = len(ub)
        return foff, nbytes_written
```
The comment about "this might not work" seems too prophetic, although I'm not sure overloading is the issue.  

First, there is a bug when the function `_fwrite_to_file` throws an exception.   The error handler does not set foff or nbytes_written.  When an exception is thrown from a low-level write error the caller aborts with an exception saying that at the return line at the end of the function foff has not been defined.   The reason is that when the write function throws an exception the definition of foff is bypassed.  This is the way I fixed that issue to get it to run, but it is not the right solution:
```
try:
                # This function has overloading.  this might not work
                foff = _fwrite_to_file(mspass_object, dir, dfile)
            except MsPASSError as merr:
                mspass_object.elog.log_error(merr)
                if kill_on_failure:
                    mspass_object.kill()
                return 0,0
```
Returning 0,0 is probably not the right thing to do, but was a workaround to confirm the issue was with _fwrite_to_file throwing an exception.   I don't know why that is happening as that is in the C++ code.  I'd need to run this under gdb in mixed C++/python debugging to find out for sure.   That, however, is not the point.   The caller of the _save_data_to_dfile is not handling this error condition correctly.   We need to design a controlled test writing to a file where writing fails and have this section of save_data handle that condition properly:
```

            if storage_mode == "file":
                # TODO:  be sure this can't throw an exception
                foff, nbytes = self._save_data_to_dfile(
                    mspass_object, dir, dfile, format=format
                )
                insertion_dict["dir"] = dir
                insertion_dict["dfile"] = dfile
                insertion_dict["foff"] = foff
                if format:
                    insertion_dict["nbytes"] = nbytes
                    insertion_dict["format"] = format
```
As you can see it definitely is not handling that condition now.   

Right now my example has really mysterious behavior after I applied the patch described above.  Document entries are written in the database but nothing was actually saved.   Any read will fail because there is no data in the files defined by dir and dfile even though the doc saved says there should be.   Not good.

**Comment by pavlis** [comment]
I found out the specific reason why the workflow I was running revealed this problem.  The reason for the failure shows an additional thing needed in this code.   That is, the C++ code, `_fwrite_to_file`, makes no effort to verify that the directory, dir, exist and is writable.  It simply creates a path, tries to open the file, and throw a MsPASSError exception if the open fails (an essential exception to catch and handle).  We either need to add code to the C++ function or the python function above to check for the existence of the named directory.  The code should follow common practice and automatically create that directory if possible and continue silently.   It should only throw an exception if it cannot create the directory.   That logic should precede the fopen call in the C function.  I tend to think this code should be in python as (a) it is easier and cleaner to do in python, and (b) it is easier to maintain as the python code base is not opaque like the C++ side.  

I will note the C function should perhaps also have try-catch blocks around the calls to fseek and ftell that write custom messages.   As the comment there says, however, fseek and ftell error handling is old school 1970s C and a bit confusing.  Then again that is perhaps why it really be wrapped in error handlers more consistent with python.

**Comment by pavlis** [comment]
Found yet another issue with our writer when writing the sample data as raw binary with fwrite in the C++ function `_fwrite_to_file`.   The metadata field "format" is not being handled correctly.   

The bad behavior is this.   My workflow reads data from wf_miniseed.   Documents in wf_miniseed will always have the entry `{"format" : "mseed"}` in the database.  The reader will properly load that key-value pair in all seismic data objects it creates.   The problem is that data saved with db.save_data to wf_TimeSeries or wf_Seismogram simply copy the `{"format" : "mseed"}` entry to the output.   That is ignored for gridfs input but the file reader does not handle that condition properly.  If format is defined in a document passed to the reader and file mode is set it always initiates a call to obspy's format cracking reader.   

A simple solution is a small patch.   The offending section of code is actually in the save_data method of Database.  This is the current code:
```
            if storage_mode == "file":
                # TODO:  be sure this can't throw an exception
                foff, nbytes = self._save_data_to_dfile(
                    mspass_object, dir, dfile, format=format
                )
                insertion_dict["dir"] = dir
                insertion_dict["dfile"] = dfile
                insertion_dict["foff"] = foff
                if format:
                    insertion_dict["nbytes"] = nbytes
                    insertion_dict["format"] = format
```
If that is only changed to the following the writer will create documents that won't break the file-based reader:
```
            if storage_mode == "file":
                # TODO:  be sure this can't throw an exception
                foff, nbytes = self._save_data_to_dfile(
                    mspass_object, dir, dfile, format=format
                )
                insertion_dict["dir"] = dir
                insertion_dict["dfile"] = dfile
                insertion_dict["foff"] = foff
                if format:
                    insertion_dict["nbytes"] = nbytes
                    insertion_dict["format"] = format
                else:
                    insertion_dict.pop("format",None)
                    insertion_dict.pop("nbytes",None)
```
Noting the call too pop uses an important trick.  Without the second argument a call to pop would need to be surrounded by an `if "format" in insertion_dict:` clause.  That is a simpler and easier trick I found on web tutorials.  

The above patch needs to be installed asap as the reader/writer pair is broken for file io right now when data begins life as a miniseed.   

The second part is a topic for discussion. In my opinion used nonexistence of a key-value pair to imply a particular format is a bad idea.  None normally means something is not defined and should be treated with care.  I know normal practice is to use arg=None to trigger default behavior not easily defined as a constant value, but in my opinion using None in the manner above to imply a specific format is a bad idea.  I think having an explicit format value for all data in the database would be a better approach with a tiny impact on storage.   I would recommend `{"format":"raw"}`, `{"format":"binary"},  or `{"format":"b"}` for something more compact.   I would, however, also recommend the code still handle undefined (None) as it does now and try to recover assuming a binary read if format is not set - the only current behavior.   Whether that should be done silently or log a complaint or informational error is debatable.

**Comment by pavlis** [comment]
The revisions in the merged branch #365 did not fix this problem.  In fact, in made did something worse.   Rather than fail a workflow I ran trying to save wf_TimeSeries output as files created no output at all and generate not a single error log entry.  Not good at all.

I have not looked at the revisions to have a hint of why I see this behavior but it needs to be fixed.   I'm running the same workflow as I type this saving to gridfs to make sure that isn't broken too.   Assume if I don't post another comment that worked. 

This is an important topic for our meeting tomorrow Feb 27, 2023

**Comment by pavlis** [comment]
Even worse news.  Something is completely broken in the workflow below.  Note the identical run worked yesterday before I pulled the newest container:
```
from mspasspy.algorithms.window import WindowData
from mspasspy.algorithms.signals import detrend
from mspasspy.algorithms.basic import ator,rtoa
from mspasspy.algorithms.edit import erase_metadata
from mspasspy.io.distributed import read_distributed_data
import time
import os

# here we write to this directory so we need to make sure it exists before starting
# if it doesn't exist create it
wftmp_dir="./wftmp"
dir_exists = os.path.exists(wftmp_dir)
if not dir_exists:
    os.makedirs(wftmp_dir)
data_per_partition=1000
erase_list=["net","sta","chan","loc"]
# error with undefined source_id without this query
query={"source_id" : {"$exists" : True},"channel_id" : {"$exists" : True}}
wfsize=db.wf_miniseed.count_documents({})
nwf_to_process=db.wf_miniseed.count_documents(query)
print("wf_miniseed input collection has ",wfsize," documents")
print(nwf_to_process," have channel and source cross-reference ids and will be processed")
npartitions=int(nwf_to_process/data_per_partition)
print("dask bag number of partitions used=",npartitions)
cursor = db.wf_miniseed.find(query).sort("source_id")
normalize_list=["channel","source"]
t0=time.time()
seisbag = read_distributed_data(db,cursor,normalize=normalize_list,npartitions=npartitions)
seisbag = seisbag.map(set_tref)
seisbag = seisbag.map(detrend,type='constant')
seisbag = seisbag.map(cautious_ator)
seisbag = seisbag.map(WindowData,-150.0,250.0)
seisbag = seisbag.map(rtoa)
seisbag = seisbag.map(erase_metadata,erase_list)
# this will not work.  The bug with foff needs to be fixed - work in progress by team feb 2023
#seisbag = seisbag.map(save_data_to_files,db,wftmp_dir)
seisbag = seisbag.map(db.save_data)
seisbag = seisbag.map(terminator)
scheduler_client.persist(seisbag)
result_list=seisbag
ntrue=0
nfalse=0
for x in result_list:
    if x:
        ntrue += 1
    else:
        nfalse += 1
print("Number still live=",ntrue)
print("Number killed=",nfalse)
t=time.time()
print("Time to process data set=",t-t0)
```
Noting it produced this output:
```
wf_miniseed input collection has  72512  documents
72039  have channel and source cross-reference ids and will be processed
dask bag number of partitions used= 72
Number still live= 0
Number killed= 72038
Time to process data set= 618.5349566936493
```
Showing that something in the new changes is killing all data passed through it.  What is really mysterious is the elog collection is empty.   NOT GOOD!

**Comment by wangyinz** [comment]
This has to be either the new writer or the new io module. @Aristoeu could you please take a closer look? Thank you!

**Comment by Aristoeu** [comment]
Yes, of course, I'm checking to see what is the problem.

**Comment by wangyinz** [comment]
I guess the problem should be in the new io module. Not sure why our tests didn't show any issues. This also means that we need to update our test to cover for cases like this one.

**Comment by Aristoeu** [comment]
I am sorry for the bugs, the problem is in the new io module. I forgot to set objects live and load the history, and I also missed the tests for history. I have added a new pull request #404  to fix the problems and updated the tests.

But I am not sure if the new code can pass the example of miniseed above. I have trouble loading the miniseed dataset and scheduler_client

**Comment by Aristoeu** [comment]
The fix still didn't solve the whole problem. When running the miniseed workflow above, `save_data()` is incorrect, and the error generated is more weird.

The code snippet I use:
```python
    seisbag = read_distributed_data(db,cursor,normalize=normalize_list,npartitions=npartitions)
    seisbag = seisbag.map(set_tref)
    seisbag = seisbag.map(detrend,type='constant')
    seisbag = seisbag.map(cautious_ator)
    seisbag = seisbag.map(WindowData,-150.0,250.0)
    seisbag = seisbag.map(rtoa)
    seisbag = seisbag.map(erase_metadata,erase_list)
    # this will not work.  The bug with foff needs to be fixed - work in progress by team feb 2023
    seisbag = seisbag.map(db.save_data)
    seisbag = seisbag.map(terminator)
    scheduler_client.persist(seisbag)
    result_list=seisbag
    ntrue=0
    nfalse=0
    for x in result_list:
        if x:
            ntrue += 1
        else:
            nfalse += 1
    print("Number still live=",ntrue)
    print("Number killed=",nfalse)
```
Running this code can save to file, but will get the error:
`
ServerSelectionTimeoutError: 127.0.0.1:27017: [Errno 111] Connection refused, Timeout: 30s, Topology Description: <TopologyDescription id: 64ebf1a8d521e96a2a675328, topology_type: Unknown, servers: [<ServerDescription ('127.0.0.1', 27017) server_type: Unknown, rtt: None, error=AutoReconnect('127.0.0.1:27017: [Errno 111] Connection refused')>]>
`

It should not be a problem with Mongo. If I delete the line `seisbag = seisbag.map(db.save_data)`, all running correct.

**Comment by wangyinz** [comment]
Isn't the error suggesting that it cannot connect to the mongodb server? If so, I would think that the problem is still in the mongodb setup in the new container.

**Comment by wangyinz** [comment]
The current file-save implementation addresses the failures described here. `_save_sample_data_to_file` normalizes the default to `format="binary"`, creates missing directories, checks directory/file writability, catches low-level binary write failures, logs the error, kills the affected datum, and returns before invalid `foff` metadata can be stored. Successful writes record `foff`, `nbytes`, and the actual format. Native-binary and formatted round trips are covered in `test_save_and_read_data`. Closing as completed.
